### ISSUE TITLE
Next release (20th anniversary release) tracking issue

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#4382]
Time for a new release issue, I think.

I propose a **hard** cut-off of **7th of November, 2020**, to be named the **20th Anniversary Edition** per:

> After a couple months of work I had something which worked, or so I thought, and _I posted [PfaEdit] to the web [...] on 7 November of 2000_. Within a month I had received my first bug report, and presumably had my first user. [Source](http://pfaedit.org/ff-history.html)

I propose a **tentative code freeze** on October 31.

I know, normally we don't do this, a hard cut-off date. But given the circumstances, I'd say it's warranted. That's why I'm getting the discussion started five months early.

**Comment by ctrlcctrlv** [comment]
Here's what, at minimum, needs to be done:

* Splash screen needs updating to accept PNG (or at least BMP) files. This either needs to happen via a build script, or by using George's PNG/BMP functions. I punted on this in #3754, but it's time. I really want to put "Celebrating 20 years!" on the splash screen, and this is the best way. It's not right for me to just do it the GDB way and spam the repo with a commit and a revert, each hundreds of thousands of lines long.

And here's what would be nice:

* At least one major feature. Perhaps @skef will be working on one :-) Hopefully I will be too! (COLR/CPAL!)

**Comment by MaddTheSane** [comment]
OpenType variable font support would be appreciated, but I don't see that happening within the cutoff date.

**Comment by ctrlcctrlv** [comment]
@MaddTheSane are you aware of my method for making variable fonts in FontForge? See e.g. https://github.com/ctrlcctrlv/Noto-Sans-Tagalog

I'm working on one with even more axes currently.

**Comment by ctrlcctrlv** [comment]
Just a quick update. The more I'm thinking about it, I think it makes sense to actually put the celebrating 20 years as part of the text and not necessarily in the image. That's a much simpler change and one that I have a lot more time to do.

**Comment by ReNardier** [comment]
it seems the anniversary is tomorrow, so happy birthday fontforge!

**Comment by frank-trampe** [comment]
@ctrlcctrlv, I'm getting bogged down in bugfixes. Do you have time to experiment with the splash screen text?

**Comment by skef** [comment]
I'm hoping this can be closer to same-month than same-day anniversary-wise. Be nice to close out more low hanging fruit prior to a release.

**Comment by frank-trampe** [comment]
@skef, I'm sure you've known me and @ctrlcctrlv long enough to know that we'd sooner fall on our swords than miss a date like this.

I'll be working pretty steadily for the rest of the month, and we can turn out a Christmas or New Year's release.

**Comment by jtanx** [comment]
better pull in the updated translations from crowdin then...

**Comment by ctrlcctrlv** [comment]
I got stuck on OS X build and then the election drama totally blew away my productivity. Sorry, fucked this one up.

I got some time yeah but it is a little bit complicated. Basically the splash screen is laid out in a very low level way. The width of each word is calculated and it's put in a row, when the row exceeds the size of the window it'll go on to the next row.

Except, George only implemented that for the longest line. And he did it with global static variables. So trying to move stuff around took longer than the hour I alloted a few weeks ago and I just haven't gotten back to it. 

Basically the smartest way to change it would be to split that into a function and apply it to all the lines. 

It's pretty frustrating that changing something this simple requires this much thinking as well 😦 

I got a segfault with something I thought would work and I just ran out of time on that day.

**Comment by ctrlcctrlv** [comment]
Oh, it's not the 7th yet. Yeah, I've kind of been lost in the election stuff. I think if we rush we can put something out. 

The SVG problems are enough reason to do it. Even without the anniversary, which puts us over the threshold of a release, the recurrent SVG bugs that keep coming back to the mailing list and the tracker are a good reason. It's not really the fault of users or @skef, at this point it's really on us for not having made a release in so long.

**Comment by frank-trampe** [comment]
This may be the best microcosm of classic FontForge code I've seen in a long time.

**Comment by ctrlcctrlv** [comment]
I have to go take a shower but I will come back and screw around with the splash screen code again. It should be fixable somehow. It might even make sense to just put a regular GLabel there and give up on this ridiculous code George wrote. 

I have a feeling this hasn't changed since the early PfaEdit days. When we changed the wording a while back we only change the longest line so it wasn't a problem. 

But trying to add a new line above that, and have the width applied to all the lines (because I noticed on my new system that the GitHub line overflows, I believe a combination of the font choice and HiDPI), is why that seemingly simple change took so long.

**Comment by frank-trampe** [comment]
Are there any major and easily fixable outstanding SVG issues?

**Comment by jtanx** [comment]
> Oh, it's not the 7th yet. 

![image](https://user-images.githubusercontent.com/5137410/98432498-f183b280-2112-11eb-939e-52b78fa36756.png)


Depends on where you live :P

**Comment by frank-trampe** [comment]
Hehe. We'll allow ourselves to run on George Williams time for this one.

**Comment by frank-trampe** [comment]
@jtanx, is there an easy way to build the Windows binaries right now or are we stuck waiting for that upstream fix?

**Comment by ctrlcctrlv** [comment]
@jtanx if you could do that, because I've never done it before, that would be appreciated. 

If you could help build on OS X and if you're really feeling generous Windows that would _really_ be appreciated. 

Or maybe @frank-trampe knows how idk. I worked a little bit with Mac stuff for Qglif and let me tell you all it's the most frustrating experience I've ever had as a developer, keeping in mind that I did some Android development in my early career on the original Google Nexus phone (2016).

**Comment by ctrlcctrlv** [comment]
No the SVG issues have all been fixed in master. People keep coming here and complaining about them because we haven't cut a release. We keep saying to download the development version when obviously the right thing to do is to cut a new release. 

So I'm just saying we have two good reasons to cut a new release.

**Comment by frank-trampe** [comment]
Ah, good.

**Comment by jtanx** [comment]
I can always build it myself, or can probably find a workaround if needed.

**Comment by ctrlcctrlv** [comment]
Erm, obviously not every outstanding SVG issue has been fixed, that is practically an impossibility, I just mean that the most serious ones that are causing crashes have been fixed.

**Comment by ctrlcctrlv** [comment]
Okay let me patch the splash screen then. I kept my branch and I should be able to figure it out if I devote another hour. The first hour was spent on finding the problem and figuring out what the code did and like ten failed attempts at splitting it into a function. (It uses static pointers, surprise!)

Shower first 🚿🛁

Thanks for the reminder @skef by the way.

**Comment by skef** [comment]
I have no objections to a release soon -- quite the opposite. We should get the old SVG fixes into the release. I'm just asking about the timing details. I would prefer if we can do the usual last-few-weeks push.

**Comment by frank-trampe** [comment]
FontForge derives its popularity in part from a lack of viable alternatives, but there's a lot of nostalgia among its die-hard users. I can't disappoint them... particularly since I'm one of them. I assure you that the community benefits of rushing the release shall exceed those of a more comfortably timed one.

I'm going to sleep shortly, but @jtanx may want to tag in order to build binaries at a convenient hour. I propose release notes as follows.

> This release falls on the 20th anniversary of the first release of FontForge back in 2000. It brings a wide range of minor tweaks and bug fixes for the user interface and file format handlers and a special splash screen to commemorate the big day.

I will write something a bit more flowery and verbose for the mailing list announcement.

**Comment by jtanx** [comment]
I'm happy to cut (or not cut) a release whenever. I'm also happy to cut another release some time later if need be, it shouldn't be that big of a deal to make one.

**Comment by jtanx** [comment]
@ctrlcctrlv this should make your life much easier to generate the splash: https://github.com/fontforge/fontforge/pull/4484/files#diff-1d99d6c05f382a70ba74d71e2b47d33aba55ad1927c49e712f206e660771789e

Although I agree it would be nice if it could just directly load a PNG or something, not so keen  on these 3MB+ source files

**Comment by ctrlcctrlv** [comment]
@jtanx My concern (above) was the two commits...one to add the "20th anniversary" and one to remove it...a lot of changed lines for no good reason :\

**Comment by jtanx** [comment]
Just change the text then?

**Comment by ctrlcctrlv** [comment]
Right, as also stated above, that's what I plan to do

**Comment by skef** [comment]
As it happens I'm unusually over-subscribed this weekend (especially for Covid) and don't want to rush fixes anyway. I'll see if I can look at a few things but otherwise I won't sweat this release.

**Comment by ctrlcctrlv** [comment]
@jtanx @frank-trampe 

Unfortunately, despite putting three more hours into it, it remains broken in an important way. I'm not making any progress, and it's 3am, so I doubt I'll make more. One of you probably knows the issue anyway, I have a headache from thinking about it :disappointed: 

## Patch
https://gist.github.com/ctrlcctrlv/9b5d4c6444eb7c0dcc0ff3a973a790a2

## Screenshot
![image](https://user-images.githubusercontent.com/838783/98439478-f874db80-20a6-11eb-8053-66b2891d5667.png)

The textbox refuses to fill the remaining ⅔ of the window in the full presentation of the preview (Help→About).

If you make the window the full size to begin with, and not the small size, though, it works. (Just add a `* 3` to line 1212.)

This patch is meant to be applied to c94b9373b4c94aada45ebdd80a480bc2ab3928d7

For some reason GTextArea thinks it doesn't have all the white space on bottom. No idea why. Sorry.

**Comment by ctrlcctrlv** [comment]
With change to line 1212:

![image](https://user-images.githubusercontent.com/838783/98439704-94ebad80-20a8-11eb-86a2-3096902cc1a8.png)

This is roughly how it is supposed to look.

Something is messed up in GDraw somewhere when the window gets resized. The image is fooling it somehow into thinking there's no space. I don't know why

**Comment by frank-trampe** [comment]
Sometimes a different set of eyes can find an easy solution to a stubborn problem, but that was not guaranteed, and we are on a tight schedule, so I instead made a new splash image in #4486. Please test that branch.

**Comment by frank-trampe** [comment]
@jtanx, I have tagged the pre-release and will await packages.

**Comment by ctrlcctrlv** [comment]
It's fine. I understand. GDraw is endless frustration. Should we want to change the text again my patch is a good place to start.

**Comment by jtanx** [comment]
@frank-trampe I'm going to shift the tag because we kind of need #4484

**Comment by jtanx** [comment]
@frank-trampe I've pushed the binaries.

**Comment by frank-trampe** [comment]
Oops. I thought for some reason that we had already merged #4484.

**Comment by frank-trampe** [comment]
I just sent the announcement to the mailing list. Cheers to good teamwork and a smooth release.

**Comment by JoesCat** [comment]
> I got stuck on OS X build and then the election drama

Yeah, was certainly something.

A bit late now, but in terms of the future, has anyone tested importing an image with GIMP and exporting as xpm ?

**Comment by ctrlcctrlv** [comment]
FYI tweeted : https://twitter.com/FontForge/status/1325288961661562881?s=19
